core/src/avm2/domain.rs

```diff
@@ -303,10 +319,25 @@ impl<'gc> Domain<'gc> {
 
     pub fn set_domain_memory(
         &self,
-        mc: MutationContext<'gc, '_>,
-        domain_memory: ByteArrayObject<'gc>,
-    ) {
-        self.0.write(mc).domain_memory = Some(domain_memory)
+        activation: &mut Activation<'_, 'gc>,
+        domain_memory: Option<ByteArrayObject<'gc>>,
+    ) -> Result<(), Error<'gc>> {
+        let mut write = self.0.write(activation.context.gc_context);
+        if let Some(domain_memory) = domain_memory {
```

Comment:
later `.unwrap_or_else` kinda duplicates this condition, so a bit simpler I think:
```rs
let new_memory = if let Some(domain_memory) = domain_memory {
    // check length
    domain_memory
} else {
    write.default_domain_memory.expect("Default domain memory not initialized")
}
write.domain_memory = Some(new_memory);